barretenberg/cpp/src/barretenberg/honk/pcs/zeromorph/zeromorph.hpp

```diff
@@ -0,0 +1,357 @@
+                                                                           Fr x_challenge,
+                                                                           Fr z_challenge)
+    {
+        // We cannot commit to polynomials with size > N_max
+        size_t N = zeta_x.size();
+        ASSERT(N <= N_max);
+
+        // Compute q_{\zeta} and q_Z in place
+        zeta_x.factor_roots(x_challenge);
+        Z_x.factor_roots(x_challenge);
+
+        // Compute batched quotient q_{\zeta} + z*q_Z
+        auto batched_quotient = zeta_x;
+        batched_quotient.add_scaled(Z_x, z_challenge);
+
+        // TODO(#742): To complete the degree check, we need to commit to (q_{\zeta} + z*q_Z)*X^{N_max - N - 1}.
+        // Verification then requires a pairing check similar to the standard KZG check but with [1]_2 replaced by
+        // [X^{N_max - N -1}]_2. Two issues: A) we do not have an SRS with these G2 elements (so need to generate a fake
+        // setup until we can do the real thing), and B) its not clear to me how to update our pairing algorithms to do
+        // this type of pairing. For now, simply construct q_{\zeta} + z*q_Z without the shift and do a standard KZG
+        // pairing check. When we're ready, all we have to do to make this fully legit is commit to the shift here and
+        // update the pairing check accordingly. Note: When this is implemented properly, it doesnt make sense to store
+        // the (massive) shifted polynomial of size N_max. Ideally would only store the unshifted version and just
+        // compute the shifted commitment directly via a new method.
+        auto batched_shifted_quotient = batched_quotient;
+
+        return batched_shifted_quotient;
+    }
+};
+
+/**
+ * @brief Verifier for ZeroMorph multilinear PCS
+ *
+ * @tparam Curve
+ */
+template <typename Curve> class ZeroMorphVerifier_ {
+    using Fr = typename Curve::ScalarField;
+    using Commitment = typename Curve::AffineElement;
+
+  public:
+    /**
+     * @brief Compute commitment to partially evaluated batched lifted degree quotient identity
+     * @details Compute commitment C_{\zeta_x} = [\zeta_x]_1 using homomorphicity:
+     *
+     *  C_{\zeta_x} = [q]_1 - \sum_k y^k * x^{N - d_k - 1} * [q_k]_1
+     *
+     * @param C_q Commitment to batched lifted degree quotient
+     * @param C_q_k Commitments to quotients q_k
+     * @param y_challenge
+     * @param x_challenge
+     * @return Commitment
+     */
+    static Commitment compute_C_zeta_x(Commitment C_q, std::vector<Commitment>& C_q_k, Fr y_challenge, Fr x_challenge)
+    {
+        size_t log_N = C_q_k.size();
+        size_t N = 1 << log_N;
+
+        auto result = C_q;
+        for (size_t k = 0; k < log_N; ++k) {
+            auto deg_k = static_cast<size_t>((1 << k) - 1);
+            // Compute scalar y^k * x^{N - deg_k - 1}
+            auto scalar = y_challenge.pow(k);
+            scalar *= x_challenge.pow(N - deg_k - 1);
+            scalar *= Fr(-1);
```

Comment:
The subtraction operator is actually not defined in this case, not entirely sure why. Possibly because these are affine points and we generally want to do operations in projective form? In any case, this will be changed in the follow on to be computed as a batch mul (since that's what's natural for the recursive computation and that's the one we care about)

Reply:
Oh right, I forgot to think about affine and projective while reviewing the PR wops. Can you add a TODO + issue for adding batch mul and generally converting point operations to projective form?

Reply:
This will all be handled in the immediate follow on but it's really not all that important anyway since the native verifier is basically nothing more than a debugging tool